src/src/PluginStructs/P105_data_struct.cpp

```diff
@@ -1,262 +1,191 @@
-      return false;
-    }
-    state            = Initialized;
-    last_measurement = 0;
-  }
+bool AHTx_Device::softReset() {
+  return I2C_write8(i2cAddress, 0xBA); // soft reset takes less than 20 msec
+}
 
-  if (state != Wait_for_samples) {
-    last_measurement = current_time;
+uint8_t AHTx_Device::readStatus() {
+  return I2C_read8(i2cAddress, NULL);
+}
 
-    startMeasurement();
-    state = Wait_for_samples;
+bool AHTx_Device::readData() {
+  const uint8_t data_len = 6;
+
+  // I2C_read8 len
+  if (Wire.requestFrom(i2cAddress, data_len) == 0) {
     return false;
   }
 
-  // make sure we wait for the measurement to complete
-  if (!timeOutReached(last_measurement + AHT10_MEASURMENT_DELAY)) {
-    return false;
+  uint8_t data[data_len];
+
+  for (uint8_t i = 0; i < data_len; ++i) {
+    data[i] = Wire.read();
   }
 
-  if (!readMeasurement()) {
-    state = Initialized;
+  // check status
+  AHTx_Status status = data[0];
+
+  if (!(status.valid() && status.calibrated())) {
     return false;
   }
 
-  last_measurement = current_time;
-  state            = New_values;
-  last_temp_val    = readTemperature();
-  last_hum_val     = readHumidity();
-
-
-  /*
-     String log;
-
-     if (loglevelActiveFor(LOG_LEVEL_INFO)) {
-     log.reserve(120); // Prevent re-allocation
-     log  = getDeviceName();
-     log += F(":");
-     }
-     boolean logAdded = false;
-
-
-     if ((tempOffset > 0.1f) || (tempOffset < -0.1f)) {
-     // There is some offset to apply.
-     if (loglevelActiveFor(LOG_LEVEL_INFO)) {
-      log += F(" Apply temp offset ");
-      log += tempOffset;
-      log += F("C");
-     }
-
-     if (hasHumidity()) {
-      if (loglevelActiveFor(LOG_LEVEL_INFO)) {
-        log += F(" humidity ");
-        log += last_hum_val;
-      }
-      last_hum_val = compute_humidity_from_dewpoint(last_temp_val + tempOffset, last_dew_temp_val);
-
-      if (loglevelActiveFor(LOG_LEVEL_INFO)) {
-        log += F("% => ");
-        log += last_hum_val;
-        log += F("%");
-      }
-     } else {
-      last_hum_val = 0.0f;
-     }
-
-     if (loglevelActiveFor(LOG_LEVEL_INFO)) {
-      log += F(" temperature ");
-      log += last_temp_val;
-     }
-     last_temp_val = last_temp_val + tempOffset;
-
-     if (loglevelActiveFor(LOG_LEVEL_INFO)) {
-      log     += F("C => ");
-      log     += last_temp_val;
-      log     += F("C");
-      logAdded = true;
-     }
-     }
-
-     if (hasHumidity()) {
-     if (loglevelActiveFor(LOG_LEVEL_INFO)) {
-      log     += F(" dew point ");
-      log     += last_dew_temp_val;
-      log     += F("C");
-      logAdded = true;
-     }
-     }
-
-     if (logAdded && loglevelActiveFor(LOG_LEVEL_INFO)) {
-     addLog(LOG_LEVEL_INFO, log);
-     }
-
-   */
+  // 20 bits humidity value
+  uint32_t value = data[1];
 
-  return true;
-}
+  value        = (value << 8) | data[2];
+  value        = (value << 4) | (data[3] >> 4);
+  last_hum_val = static_cast<float>(value) / (1 << 20) * 100;
```

Comment:
Same here... do you mean to divide it by roughly 100 million, or divide by a million and multiply by 100 ?

Reply:
That's (all) inherited code... only changed the casting from the original Playground code...

Will fix after walking the dog 🐕

Reply:
The formulas provided in the data sheet look like this:

![Screenshot - 01_08_2021 , 20_06_29](https://user-images.githubusercontent.com/1922393/127781039-5c5bcbb7-3ade-4446-95a6-02c678287c5a.png)